src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateScheduler.scala

```diff
@@ -0,0 +1,451 @@
+              "Critical error. Missing parent " +
+                s"with hash ${parent}"
+            )
+          )
+          val newParentDeps = parentRequest.dependencies - 1
+          if (newParentDeps == 0) {
+            // we can always call `parentRequest.resolvedData.get` on parent node, as to even have children parent data
+            // needs to be provided
+            go(
+              currentRequests - parent,
+              currentBatch + (parent -> (parentRequest.resolvedData.getOrElse(
+                throw new IllegalStateException(
+                  s"Critical error. Parent ${parentRequest.nodeHash} without resolved data"
+                )
+              ), parentRequest.requestType)),
+              parentsToCheck.tail ++ parentRequest.parents
+            )
+          } else {
+            go(
+              currentRequests + (parent -> parentRequest.copy(dependencies = newParentDeps)),
+              currentBatch,
+              parentsToCheck.tail
+            )
+          }
+        }
+      }
+
+      val newActive = activeRequest - request.nodeHash
+      val newMemBatch = memBatch + (request.nodeHash -> (request.resolvedData.get, request.requestType))
+
+      val (newRequests, newBatch) = go(newActive, newMemBatch, request.parents)
+      copy(activeRequest = newRequests, memBatch = newBatch)
+    }
+
+    def resolveRequest(
+        request: StateNodeRequest,
+        receivedData: ByteString,
+        requestNewChildren: Seq[StateNodeRequest]
+    ): SchedulerState = {
+      val numberOfChildren = requestNewChildren.size
+      val resolvedStateNodeRequest =
+        request.copy(resolvedData = Some(receivedData), dependencies = request.dependencies + numberOfChildren)
+      val newRequests = activeRequest + (request.nodeHash -> resolvedStateNodeRequest)
+      val stateWithUpdatedParent = copy(activeRequest = newRequests)
+      requestNewChildren.foldLeft(stateWithUpdatedParent) { case (state, child) => state.schedule(child) }
+    }
+
+    def getNodesToPersist: (Seq[(ByteString, (ByteString, RequestType))], SchedulerState) = {
+      (memBatch.toSeq, copy(memBatch = Map.empty))
+    }
+  }
+
+  object SchedulerState {
+    def apply(): SchedulerState = {
+      new SchedulerState(
+        Map.empty[ByteString, StateNodeRequest],
+        PriorityQueue.empty(stateNodeRequestComparator),
+        Map.empty
+      )
+    }
+  }
+
+  case object ProcessingSuccess
+
+  sealed trait ResponseProcessingError
+
+  sealed trait CriticalError extends ResponseProcessingError
+
+  case object CannotDecodeMptNode extends CriticalError
+
+  case object NotAccountLeafNode extends CriticalError
+
+  sealed trait NotCriticalError extends ResponseProcessingError
+
+  case object NotRequestedItem extends NotCriticalError
+
+  case object AlreadyProcessedItem extends NotCriticalError
+
+  final case class ProcessingStatistics(duplicatedHashes: Long, notRequestedHashes: Long, saved: Long) {
```

Comment:
Do we want to expose ProcessingStatistics as a metric?

Reply:
it is possible, we can even save it to do to save stats between shutdowns. I will probably do this when sync will be ready for its prime time(i.e it will work with mainnet)